### ISSUE TITLE
Appropriate component baseline

### ISSUE THREAD

**Issue by mgarin** [mgarin/weblaf#424]
It is a major internal enhancement I've been working on in parallel with other changes, so I decided to create this separate issue where I will post all updates regarding the progress made. So let me describe how what is it, how things have been working so far and what changes are coming...

### Component baseline

Some of existing Swing components are able to provide a setting called "baseline". It is basically an Y coordinate (counting from the top) within component bounds which defines, well, a base line of the component. That coordinate is used by multiple Swing and custom layouts to align components by their baseline, it is most useful for aligning component that have text part within.

Here is a small example that displays three components aligned by their baseline:
![image](https://cloud.githubusercontent.com/assets/2361389/20630423/e9361962-b341-11e6-9ec9-18ad609d47fa.png)

As you can see - each component has its own different height/border, but their baseline which is calculated based on the inner text perfectly matches within the line. 

So basically it makes the lined up components look much better because it is much more convenient to read text that is perfectly lined up rather than a text that jumps up and down in each component.

### How it works

In Swing all baseline calculations are hardcoded within each of the component UIs - for example you can find such code within `BasicLabelUI` or `BasicButtonUI` classes. Even though that code might be quite similar in many of the existing UIs it is simply copy-pasted everwhere. As well as the text painting code actually which includes painting plain and HTML text.

I'm not saying that those calculations are wrong - no, they are correct. But they are extremely hard to modify and maintain. Any changes made to the UI might cause the baseline to return wrong values ( which btw already happened once - #344 ).

### Changes coming

If you have been following the `styling` branch development process you might know that I've successfully broken down most of the existing components painting into small pieces - decorations and contents. That change have been out for a while now and it allows WebLaF users to customize any component without getting stuck with old nasty UI code or even without touching code at all.

That being said, baseline calculation haven't been updated with that change and was still performed within basic UI classes for each of the existing components. Those are correct with the default styles, but might become invalid with some deeper style customizations made.

So I've been working on a major change that would allow me to get rid of most basic UI classes ( which are only kept for the sake of having proper baseline ) and move to a new fully implemented system of baseline calculation based on component contents.

In terms of WebLaF that means that each `IContent` implementation will be able to say whether or not it can provide its own baseline and will be able to provide it as well as the baseline behavior. Obviously implementations like `AbstractTextContent` will have appropriate baseline calculations by default and will be providing it instead of the basic UI classes to any component that uses it.

This is fully an internal library change that should have no impact on any existing styles or code as usually you do not need to work with baseline directly anywhere, you only use layouts that request baseline to align components appropriately.

### Why it is major

First of all it is major because it will fully replace one of the basic features of Swing components that have been there for ages with my own implementation which will give more customization options and fix some of the old issues Swing implementation has.

It is also major because along with this change comes a necessity to fix multiple minor internal issues that have been crippling the code for a while and causing some problems with UI, painter, decoration and content implementations.

And it is major because along with that change I will be getting rid of multiple basic UI class usages and replacing them with my own implementations that will only contain code that is reasonable for WebLaF UI implementations. That will ensure that I don't have to put weird workarounds all over the UI classes code and that it doesn't perform unnecessary updates like it was with the basic progress bar UI.

**Comment by mgarin** [comment]
These changes are now available in `styling` branch sources and builds based on it since ( 45c60d8 ) commit.
